File: Lab 9 (17 April 2025)/diff colour adjacent nodes m colours n nodes.cpp
#include <iostream>
#include <vector>
using namespace std;

bool isSafe(int node, vector<vector<int>>& graph, vector<int>& color, int n, int col)
{
    for (int k = 0; k < n; ++k)
    {
        if (graph[node][k] && color[k] == col)
            return false;
    }
    return true;
}

bool solve(int node, vector<vector<int>>& graph, vector<int>& color, int m, int n)
{
    if (node == n) return true;

    for (int col = 1; col <= m; ++col)
    {
        if (isSafe(node, graph, color, n, col))
        {
            color[node] = col;
            if (solve(node + 1, graph, color, m, n))
                return true;
            color[node] = 0;
        }
    }
    return false;
}

bool graphColoring(vector<vector<int>>& graph, int m, int n)
{
    vector<int> color(n, 0);

    if (solve(0, graph, color, m, n))
    {
        cout << "Solution Exists: Node Colors are -> ";
        for (int c : color)
            cout << char('a' + c - 1) << " ";
        cout << endl;
        return true;
    }
    else
    {
        cout << "No solution exists.\n";
        return false;
    }
}

int main()
{
    int n, m;
    cout << "Enter number of nodes: ";
    cin >> n;

    cout << "Enter number of colors: ";
    cin >> m;

    vector<vector<int>> graph(n, vector<int>(n));
    cout << "Enter adjacency matrix:\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> graph[i][j];

    graphColoring(graph, m, n);

    return 0;
}

File: Lab 9 (17 April 2025)/diff colour adjacent nodes m colours n nodes.py
import time
import random
import matplotlib.pyplot as plt

def is_safe(node, graph, color, c):
    for i in range(len(graph)):
        if graph[node][i] == 1 and color[i] == c:
            return False
    return True

def solve(node, graph, m, color):
    if node == len(graph):
        return True
    for c in range(1, m + 1):
        if is_safe(node, graph, color, c):
            color[node] = c
            if solve(node + 1, graph, m, color):
                return True
            color[node] = 0
    return False

def graph_coloring(graph, m):
    n = len(graph)
    color = [0] * n
    return solve(0, graph, m, color)

def generate_random_graph(n, edge_prob=0.5):
    graph = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(i + 1, n):
            if random.random() < edge_prob:
                graph[i][j] = graph[j][i] = 1
    return graph

def measure_execution_time(m=3):
    sizes = list(range(2, 13))
    times = []
    for n in sizes:
        graph = generate_random_graph(n, edge_prob=0.5)
        start = time.time()
        graph_coloring(graph, m)
        end = time.time()
        times.append(end - start)
    return sizes, times

sizes, times = measure_execution_time()

plt.plot(sizes, times, marker='o', color='red')
plt.title("Time Complexity of M-Coloring (Backtracking)")
plt.xlabel("Number of Nodes (n)")
plt.ylabel("Execution Time (seconds)")
plt.grid(True)
plt.show()